pkg/dns/dns.go

```diff
@@ -268,6 +301,97 @@ func (kd *KubeDNS) updateService(oldObj, newObj interface{}) {
 	kd.newService(newObj)
 }
 
+func assertIsIngress(obj interface{}) (*ext.Ingress, bool) {
+	if ing, ok := obj.(*ext.Ingress); ok {
+		return ing, ok
+	} else {
+		glog.Errorf("Type assertion failed! Expected object of type \"Ingress\", got %T", obj)
+		return nil, ok
+	}
+}
+
+func (kd *KubeDNS) newIngress(obj interface{}) {
+	if ing, ok := assertIsIngress(obj); ok {
+		glog.V(4).Infof("Add/Update for ingress %v", ing.Name)
+		subCache := NewTreeCache()
+		subCachePath := append(kd.domainPath, ingressSubdomain, ing.Namespace)
+
+		ips := []string{}
+		ingFQDN := strings.Join([]string{ing.Name, ing.Namespace, ingressSubdomain, kd.domain}, ".")
+		reverseRecord, _ := getSkyMsg(ingFQDN, 0)
+		for _, lbi := range ing.Status.LoadBalancer.Ingress {
+			// We generate an A record if the Ingress loadbalancer IP is set and a
+			// CNAME record if the hostname is set. If both IP and hostname fields
+			// are set we favor IP over the hostname.
```

Comment:
DNS spec doesn't allow mixing up A and CNAME records for a given domain name. So we have to pick one or the other. And IP address is more specific than a CNAME because CNAME should ultimately resolve to some IP address, so choosing this order.